### PR TITLE
Work Package 1

### PR DESCRIPTION
**Pull request by jack89roberts**
This PR will be merged once the first milestone has been completed, see https://github.com/alan-turing-institute/ARC-LoCoMoSeT/milestone/1

This is loosely WP1 + part of WP2 as envisioned in the original statement of work.

### CODE REVIEW

**Comment by jack89roberts** [comment]
Remaining TODOs:

- [x] Finish tests. I got to creating a dummy dataset, model, preprocessor today but not incorporated them into tests for the metric running functions (and their precursors).
- [x] ~~Tidy up notebooks.~~ - **_not as part of this work package_**.
- [x] Save the preliminary results and plots somewhere.
- [x] Check documentation in general.

Then I think this is fine to merge and call WP1 done.

**Comment by jack89roberts** [comment]
TODOs:

Review (@lannelin ):

- Check the usage of HuggingFace models, datasets etc. is sensible
- General package/code structure
- Check it runs/the instructions work (on Oxford Pets probably as ImageNet takes so long - e.g. see `configs/config_pets.yaml`)

Review (@eddableheath ):

- Look over updates to tests
- Look over initial PARC results (see slides/files on SharePoint - bookmark in the Slack channel) and double-check/debug anything in the PARC implementation.

Development TODOs (@jack89roberts):

- [x] Documentation for plotting scripts (what the plots show/mean)
- [x] Consider where plot functions, metric run script, imagenet-1k scores library should be (in/out of the package structure)
- [x] Remove the notebooks from this PR unless we're sure they add something for WP1
- [x] ~~Double-check PARC implementation.~~ - WP2
- [x] Make issues for anything outstanding to be checked/improved in later work packages, e.g.
  - choice of batch size for preprocessing/feature extraction
  - caching options (need to empty the cache regularly to avoid filling the disk with processed datasets for each model used)

File: src/locomoset/datasets/preprocess.py
Comment (lannelin) [inline src/locomoset/datasets/preprocess.py]: note on expected format for sample? can be in docstring if typing is inappropriate due to mixed dict

File: src/locomoset/metrics/renggli.py
Comment (lannelin) [inline src/locomoset/metrics/renggli.py]: careful of mutable default args (https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments). I *think* in this case the params of the classifier will just get overwritten on the next `fit` call but that's always the case.
I'd suggest having this as an optional param and moving the initialisation into the fn body when `None`

File: tests/data/README.md
Comment (lannelin) [inline tests/data/README.md]: Neat. May be worth noting that these are present in the repo by default so would be a regeneration to run again.

File: tests/test_parc.py
Comment (lannelin) [inline tests/test_parc.py]: what's the expected impact of this, what should I be taking from this note?

File: tests/test_renggli.py
Comment (lannelin) [inline tests/test_renggli.py:26]: given tolerance is quite high here (compared to default values in testing libs), could we note why we think this is appropriate

File: tests/test_run_metrics.py
Comment (lannelin) [inline tests/test_run_metrics.py:16]: is it the case that *only* string values can be specified without being contained in a list? it seems intuitive that I should be able to specify
```
my_config = {
        "name_1": "name_1",
        "name_2": ["name_21", "name_22"],
        "par_1": [1, 2, 3],
        "par_2": [4, 5, 6],
        "par_3": 7
    }
``` 
where I expect `"par_3"` to be constant across all runs

Comment (lannelin) [reply tests/test_run_metrics.py:16]: in which case, I think it's a case of simply refactoring `metrics.run.nest_string_in_list` to box all non-iterable-excluding-strings in a list. That should work with the existing (diabolical) ` product(*list(map(` :grimacing:

i.e. `[x] if not isinstance(x, Iterable) or isinstance(x, str) else x`

File: src/locomoset/models/features.py
Comment (lannelin) [inline src/locomoset/models/features.py:26]: this is non-conventional afaik. You'd usually see the dataset wrapped up with a dataloader and then a loop through that. For training that makes sense (to pass into a training step) but for inference I suppose we have more flexibility and working directly with a dataset map could be an option.
It might be worth pulling out lambda as an inference step fn so you can control devices (run on gpu) per batch. Sorry if I've missed this if handled elsewhere.

File: scripts/plot_vs_actual.py
Comment (lannelin) [inline scripts/plot_vs_actual.py]: some overlap with `plot_vs_samples.py`, not too worried so up to you both

File: README.md
Comment (lannelin) [inline README.md]: typo "patch"->"path"

**Review by lannelin — COMMENTED** [review]
looking great! A few minor comments beyond the points you've already covered above.

Linked to one of my comments, you might want to consider how this would run on GPU. Both a dependencies q (i.e. ensure torch with cuda) and ensuring this happens in the code. I've seen `accelerator` (https://huggingface.co/docs/accelerate/index) used to abstract this away recently. Very happy to leave this to next work package.

I should stress that I **haven't** reviewed this from a paper implementation perspective i.e. I haven't inspected your implementation vs the paper as I think your combined expertise outweighs mine here and I imagine you've covered this between you

I don't think any of my points absolutely require changes so let me know your thoughts on the points I've raised and I can approve after discussion.

File: tests/test_run_metrics.py
Comment (eddableheath) [reply tests/test_run_metrics.py:16]: in the example you've given the itertools function would treat "name_1" as an iterable and sweep over the characters of the string for each parameter set, hence requiring any string be wrapped in a list so it gets picked up as a single element of an iterable rather than an iterable itself. 

The "par_3" value would remain constant in this case.

File: tests/test_parc.py
Comment (eddableheath) [reply tests/test_parc.py]: Updated the docstring with justification.

**Review by eddableheath — APPROVED** [review]
All looks good to me, happy with where it's at.

File: src/locomoset/models/features.py
Comment (jack89roberts) [reply src/locomoset/models/features.py:26]: Can chat about this one when you're back but is it the use of map here you think is unconventional or the way we've wrapped this into the parent `get_features` function? I think it follows how I see HuggingFace datasets used in general, e.g. https://huggingface.co/docs/datasets/image_process .

Agree it may turn out we need to be able to control some more parameters for performance etc. here later, but I don't think that needs to be part of this PR.

File: tests/test_parc.py
Comment (jack89roberts) [reply tests/test_parc.py]: Edmund's expanded the note

File: tests/test_run_metrics.py
Comment (jack89roberts) [reply tests/test_run_metrics.py:16]: @eddableheath can respond here but my understanding is it already works as written for other single, non-iterable things, the problem is just the edge-case that a string is treated as an iterable of characters rather than a single entity.

File: README.md
Comment (jack89roberts) [reply README.md]: fixed

File: tests/data/README.md
Comment (jack89roberts) [reply tests/data/README.md]: done

File: tests/test_run_metrics.py
Comment (lannelin) [reply tests/test_run_metrics.py:16]: @jack89roberts I think it should give an error `TypeError: 'int' object is not iterable` if you try the test with the `"par_3"` param included.
@eddableheath I think the logic should work? 
```
>>> from typing import Iterable
>>> maybe_box = lambda x: [x] if not isinstance(x, Iterable) or isinstance(x, str) else x
>>> maybe_box("hello")
['hello']
>>> maybe_box(["hello"])
['hello']
>>> maybe_box(1)
[1]
>>> maybe_box([1])
[1]
```

Comment (lannelin) [reply tests/test_run_metrics.py:16]: the above fn would replace `nest_string_in_list` though perhaps not as a lambda!

File: src/locomoset/models/features.py
Comment (lannelin) [reply src/locomoset/models/features.py:26]: ah sorry, non-conventional for passing anything through the network - typically see this for preprocessing data (I guess you could think of feature extraction as this??). Agree not necessary to change in this PR but wanted to flag - should have been clearer, sorry!

File: tests/test_run_metrics.py
Comment (eddableheath) [reply tests/test_run_metrics.py:16]: yes, you're completely correct, will update the function and the tests

Comment (eddableheath) [reply tests/test_run_metrics.py:16]: fixed with this commit [14e3317](https://github.com/alan-turing-institute/ARC-LoCoMoSeT/pull/15/commits/14e331740b402c7ff268a55261448580312820a2)

File: src/locomoset/models/features.py
Comment (jack89roberts) [reply src/locomoset/models/features.py:26]: Ah yeah because this is inference rather than preprocessing basically, I see what you mean

File: src/locomoset/datasets/preprocess.py
Comment (jack89roberts) [reply src/locomoset/datasets/preprocess.py]: done

File: .vscode/settings.json
Comment (lannelin) [inline .vscode/settings.json:1]: did you mean to add this file? if not, might be worth adding the folder to gitignore

File: src/locomoset/models/features.py
Comment (jack89roberts) [reply src/locomoset/models/features.py:26]: I'm going to make an issue about this rather than doing it here I think because it's not a super quick fix.

File: tests/test_run_metrics.py
Comment (lannelin) [inline tests/test_run_metrics.py]: to be removed. Would also highly recommend running individual tests via the testing GUI in vscode (let's you debug!) rather than manually like this.

**Review by lannelin — APPROVED** [review]
a couple of outstanding points from last round but I don't think anything breaking for this WP. Happy to approve. Excellent work, both of you.

File: src/locomoset/metrics/renggli.py
Comment (jack89roberts) [reply src/locomoset/metrics/renggli.py]: done

File: tests/test_run_metrics.py
Comment (jack89roberts) [reply tests/test_run_metrics.py]: done

File: .vscode/settings.json
Comment (jack89roberts) [reply .vscode/settings.json:1]: done

File: tests/test_renggli.py
Comment (jack89roberts) [reply tests/test_renggli.py:26]: done

File: src/locomoset/models/features.py
Comment (jack89roberts) [reply src/locomoset/models/features.py:26]: See #32

**Pull request merged**
